### ISSUE TITLE
Object.clone() should preserve an array's length annotations

### ISSUE THREAD

**Issue by mernst** [kelloggm/checker-framework#146]
Calling `myArray.clone()` should return an array with the same length annotations.
Below is an example of code that I think should type-check.

```java
import java.util.Arrays;
import org.checkerframework.checker.index.qual.SameLen;

public class LengthOfOtherArray {

  public static void main(String[] args) {
    String @SameLen({"args", "args_sorted"}) [] args_sorted = args.clone();
    Arrays.sort(args_sorted);
    String @SameLen({"args", "args_sorted"}) [] args_sorted2 = args_sorted.clone();
    if (args_sorted.length == 1) {
      System.out.println(args[0]);
    }

  }

} 
```

Changing the Index Checker's annotated JDK like this didn't have any effect on the errors issued:
```
-    protected native Object clone() throws CloneNotSupportedException;
+    protected native @PolyMinLen @PolySameLen Object clone(@PolyMinLen @PolySameLen Object this) throws CloneNotSupportedException;
```

**Comment by smillst** [comment]
Note this is only a problem with the SameLen Checker. 

The code below type-checks.

```
String[] argsClone = {"a"};
String @ArrayLen(1) [] abc = argsClone.clone();
```

**Comment by kelloggm** [comment]
Interestingly, the problem here is not that `Object.clone()` is incorrectly annotated, but that the array actually being assigned is apparently not valid as its own `SameLen`. Here is a simplified test case:

```java
class ObjectClone {

    void test(int[] a, int @SameLen("#1") [] b) {
	int @SameLen("a") [] c = b.clone(); // no error
	int @SameLen({"a", "d"}) [] d = b.clone(); // error: assignment incompatible
	int @SameLen({"a", "e"}) [] e = b; // error is still present; clone isn't needed
    }
}
```
This is still a problem, and I'll look into fixing it.

**Comment by kelloggm** [comment]
An even simpler test case:

```java
void foo(int[] b) {
    int @SameLen("a") [] a = b;
}
```

This problem is actually quite tricky.  When visiting assignments, we collect all the reasonable `SameLen` arrays as part of flow based refinement (i.e. in `SameLenTransfer`'s `visitAssignment` function). However, assignments are also visited by the common assignment check to ensure that `SameLen` annotations are "propagated"---that is, that when the common assignment check might issue an error, it first makes sure that it has collected all of the `SameLen` annotations on both the relevant variables _and the variables listed in the rhs' `SameLen` type_. This latter step appears not to take advantage of (or even be aware of) flow-sensitive refinement when computing the type of the rhs, which results in the bug we see above.

**Comment by kelloggm** [comment]
closed by typetools#1320
